samza-core/src/main/java/org/apache/samza/operators/impl/OperatorImplGraph.java

```diff
@@ -77,24 +79,28 @@
   private final Map<SystemStream, InputOperatorImpl> inputOperators = new HashMap<>();
 
   /**
-   * A mapping from {@link JoinOperatorSpec} IDs to their two {@link PartialJoinFunction}s. Used to associate
-   * the two {@link PartialJoinOperatorImpl}s for a {@link JoinOperatorSpec} with each other since they're
+   * A mapping from {@link JoinOperatorSpec} IDs to the single copy of {@link JoinOperatorSpec} per each task, and their
+   * two {@link PartialJoinFunction}s.
+   * Used to associate the two {@link PartialJoinOperatorImpl}s for a {@link JoinOperatorSpec} with each other since they're
    * reached from different {@link OperatorSpec} during DAG traversals.
    */
-  private final Map<String, KV<PartialJoinFunction, PartialJoinFunction>> joinFunctions = new HashMap<>();
+  private final Map<String, Tuple3<JoinOperatorSpec, PartialJoinFunction, PartialJoinFunction>> joinSpecs = new HashMap<>();
 
+  /**
+   * Method to get the current time.
+   */
   private final Clock clock;
 
   /**
    * Constructs the DAG of {@link OperatorImpl}s corresponding to the the DAG of {@link OperatorSpec}s
    * in the {@code streamGraph}.
    *
-   * @param streamGraph  the {@link StreamGraphImpl} containing the logical {@link OperatorSpec} DAG
+   * @param streamGraph  the {@link SerializedStreamGraph} containing the logical {@link OperatorSpec} DAG
    * @param config  the {@link Config} required to instantiate operators
    * @param context  the {@link TaskContext} required to instantiate operators
    * @param clock  the {@link Clock} to get current time
    */
-  public OperatorImplGraph(StreamGraphImpl streamGraph, Config config, TaskContext context, Clock clock) {
+  public OperatorImplGraph(final SerializedStreamGraph streamGraph, final Config config, final TaskContext context, final Clock clock) {
```

Comment:
+1, SerializedStreamGraph feels odd since it's not actually serialized anymore, and the name seems to describe the impl than the purpose.

Another alternative for naming/structuring this might be to have a StreamGraphSpec and StreamGraphImpl similar to the OperatorSpec and OperatorImpl duality. I.e., the existing StreamGraphImpl would then become the StreamGraphSpec that's serialized, and the deserialized version used here would be StreamGraphImpl.